scipy/stats/_qmc.py

```diff
@@ -1325,3 +1326,117 @@ def reset(self):
         """
         self.engine.reset()
         return self
+
+
+class PoissonDisc(QMCEngine):
+
+    # Uniform sampling in a hypersphere
+    # Based on Matlab implementation by Roger Stafford
```

Comment:
What does it mean? I hope this is not a reference to licensed code from Matlab otherwise we would need to replace this part entirely. Same for the other references. We need to be sure that the code we copie or just look at is not under GPL or other non permissive license.

Reply:
Sad to read this, but code from observable cannot be used like this:
https://talk.observablehq.com/t/copyright-and-software-licenses/364/2

You would need to get a relicensing from the original author here.

Reply:
IIRC, it's based on this piece of MATLAB code: https://www.mathworks.com/matlabcentral/fileexchange/9443-random-points-in-an-n-dimensional-hypersphere (or a reference to it in another place on Mathworks). It's not really reusing the code, since it is already rewritten in Python, but it uses the same gamma function magic (that I don't completely understand) to make a uniform sample in a n-dim sphere.

The other references are also to the original ideas (e.g., fopr using a circle instead fo a sphere), and not to code itself.

Reply:
Ok, I did some digging, and the specific piece of code for the uniform sampling is something I've lifted from this stackoverflow answer: https://stackoverflow.com/questions/5408276/sampling-uniformly-distributed-random-points-inside-a-spherical-volume/44782884 
I can reach out to the author and ask him if it's okay to include it in SciPy, I guess?

Reply:
That code has `No License`, so you indeed cannot use it - which includes translating it to Python - without getting permission from the original author. I suggest contacting him.

Reply:
As for StackOverflow, if you click on share you have the information about the license (depending on how old it is, a variant of CC-BY-SA). But if the answer uses non-licensed/non-compatible-licensed code, the issue remains as you are not allowed to relicensed something which is not yours without consent.

Reply:
For StackOverflow, it's licenced under CC BY-SA 3.0, so we should be fine there. 

I really doubt that we can still contact the original Matlab one-liner author from 16 years ago, but I'll send him a message.

The alternative would be, I guess, to find a publication on n-dim sphere sampling and rewrite this one line of code? It's also possible to exclude the inner part of the sphere while we're at it, then it definitely will be a different code.

Reply:
I did not look yet at your code, but from Bridson, it should sample from the hypersphere between [r/2, r]. So it's a matter of sampling uniformly from this space.

But we could also just sample using a hypercube. This way it's simpler for both the generation of a new point and for discarding cells. When you have a new point, you discard immediately the hypercube [0, r] from the list of possible cells and also discard active points which are in the hypercube (you would keep a dictionary of the index along with the coordinates of the points). Then when you sample a still valid active point from the hypercube [r/2, r], you just take a point which is still in a valid cell.

If we do this, there is no copyright issue at all.

Reply:
For the sampling part, I would even propose something simpler. Instead of sampling the hypercube [r/2, r] with float values, you would sample cells. At the very end, with all the cells we add a random perturbation to all points. You can have a fairly simple implementation with just some "box picking" using this (`random.choice`).

Reply:
Sampling from a hypercube has a problem that it loses the rotational invariance w.r.t. the axes. It probably won't matter much if you sample in the parameter space and your goal is to avoid clustering, bit it will get important if you sample e.g. in a 3D Euclidean space and want the sampling to be completely isotropic (that was actually the usecase that I originally needed this sampling for).

One way to avoid this is to sample in a sphere using a hypercube and a rejection rule (for l > 2r and l < r). This, however, will be problematic in higher dimensions since most of the points will be rejected.

An option is to have two variants, I guess: a sphere-based isotropic one for low-dimensional spaces and a hypercube-based one for higher dimensions.

There is also a third option already implemented, which I personally like quite a lot: sampling on the surface of the sphere instead. This results in a denser sampling, and is also fairly easy to implement. The caveat is that it is not the original Bridson sampling, and this method does not allow us to obtain every single maximal Poisson Disc distribution (we can't get some of the sparser ones).

Reply:
> For StackOverflow, it's licenced under CC BY-SA 3.0, so we should be fine there.

No, CC-BY-SA is not compatible with scipy's license. If that's the license of the code, we cannot use it.

Reply:
@rkern why is that? Because of the SA? We for sure have lots of code coming from StackOverflow (same for a lot of other OS projects too).

Reply:
Yes, the SA part is a copyleft provision akin to the GPL.

Reply:
It was discussed during the review and it was validated. We had some other recent PR with SO code (don't remember now which one exactly)

cc @rgommers

Reply:
Can you find the specific discussion? Github search (and search-in-page) is not helping me locate the comments in that PR.

Reply:
> Sampling from a hypercube has a problem that it loses the rotational invariance w.r.t. the axes. It probably won't matter much if you sample in the parameter space and your goal is to avoid clustering, bit it will get important if you sample e.g. in a 3D Euclidean space and want the sampling to be completely isotropic (that was actually the usecase that I originally needed this sampling for).

@diregoblin In any case, the constraint [r, 2r] is arbitrary, so I don't think it would be a problem. 

> One way to avoid this is to sample in a sphere using a hypercube and a rejection rule (for l > 2r and l < r). This, however, will be problematic in higher dimensions since most of the points will be rejected.

The rejection is not feasible in high dimensions, as you said. We should be able to generate good samples (in terms of space coverage) in high dimensions in a matter of seconds at most. I tested your code (not this PR yet, previous links you sent in the other issue) and it's already quite slow in 4-5 dimensions so we might want to do things differently.

> An option is to have two variants, I guess: a sphere-based isotropic one for low-dimensional spaces and a hypercube-based one for higher dimensions.

Yes and no. It's an interesting approach, but the two approaches need a slightly different logic for the rejection/selection. So it might be hard to develop and maintain.

> There is also a third option already implemented, which I personally like quite a lot: sampling on the surface of the sphere instead. This results in a denser sampling, and is also fairly easy to implement. The caveat is that it is not the original Bridson sampling, and this method does not allow us to obtain every single maximal Poisson Disc distribution (we can't get some of the sparser ones).

Sampling only on the surface is very dangerous in high dimensions in terms of space coverage. I am not a fan here.

Reply:
@rkern See here: https://github.com/scipy/scipy/pull/10844/files?file-filters%5B%5D=.npz&file-filters%5B%5D=.pyi&file-filters%5B%5D=.pyx&file-filters%5B%5D=.rst&file-filters%5B%5D=dotfile#r558889182

Reply:
> the constraint [r, 2r] is arbitrary, so I don't think it would be a problem
What I'm saying is, in the hypercube approach the constraint is [r, 2r] _along each axis_, while in the hypersphere approach it is [r, 2r] _for the L2 norm_. I'm pretty sure that only the latter approach results in an isotropic distribution indepentent from the coordinate system in e.g. 3D Euclidean space,

If the aim here is to sample in a heterogeneous parameter space, that's a different matter. If it makes no sense to rotate the coordinate system anyway, and our aim is to sample the space randomly and avoid clustering, the hypercube approach is feasible, I think (it should be made clear in the documentation though, to avoid misuse).

> Sampling only on the surface is very dangerous in high dimensions in terms of space coverage. I am not a fan here.

Can you elaborate on this a little bit? The surface sampling leads to a very strong peak on the radial distribution function, meaning most of the points are spaced (almost) at the minimal distance away from their neighbours. This means we exclude quite a lot of combinations of points. Each single point in the space can still be sampled though, since the 1st is drawn randomly.

> I tested your code (not this PR yet, previous links you sent in the other issue) and it's already quite slow in 4-5 dimensions so we might want to do things differently.

Yeah, it gets rather slow in higher dimensions. The main issue here is checking the neighbourhood of the candidate point for existing points, since the size of this neighbourhood grows exponentially with the number of dimensions. One possible approach is to use k-d trees instead for keeping track of the existing points, that should work well in higher dimensions. However, implementing it is quite a task by itself. I offered it as a BSc graduation project a couple of months back, but I couldn't find a suitable student.

Reply:
> What I'm saying is, in the hypercube approach the constraint is [r, 2r] _along each axis_, while in the hypersphere approach it is [r, 2r] _for the L2 norm_. I'm pretty sure that only the latter approach results in an isotropic distribution indepentent from the coordinate system in e.g. 3D Euclidean space,

True but as we sample only one point, I am not just that we would effectively notice anything. One could also still use what I propose and discard some boxes at the corners to have something close to a hypersphere. 

> If the aim here is to sample in a heterogeneous parameter space, that's a different matter. If it makes no sense to rotate the coordinate system anyway, and our aim is to sample the space randomly and avoid clustering, the hypercube approach is feasible, I think (it should be made clear in the documentation though, to avoid misuse).

The aim would be to create a hypercube which respect the minimal radius we imposed. Nothing more IMO.

> Can you elaborate on this a little bit? The surface sampling leads to a very strong peak on the radial distribution function, meaning most of the points are spaced (almost) at the minimal distance away from their neighbours. This means we exclude quite a lot of combinations of points. Each single point in the space can still be sampled though, since the 1st is drawn randomly.

Sampling on the surface means that you need to fix somehow the radius between [r, 2r], so I am worried we would bias too much here. Although might be not that different than my box proposal.

> Yeah, it gets rather slow in higher dimensions. The main issue here is checking the neighbourhood of the candidate point for existing points, since the size of this neighbourhood grows exponentially with the number of dimensions. One possible approach is to use k-d trees instead for keeping track of the existing points, that should work well in higher dimensions. However, implementing it is quite a task by itself. I offered it as a BSc graduation project a couple of months back, but I couldn't find a suitable student.

A tree approach is interesting indeed. But not sure totally needed. Even if using your approach, you should still keep track of the points in terms of coordinates and boxes, so that you can instantly get the boxes around. This is what Bridson says in his (very short) article.

Reply:
> The aim would be to create a hypercube which respect the minimal radius we imposed. Nothing more IMO.

Noted, then a hypercube is indeed a cheap alternative

> Sampling on the surface means that you need to fix somehow the radius between [r, 2r], so I am worried we would bias too much here.

In my case, I fix it to the minimal radius r, to obtain a valid PD distribution, while also picking as many points as possible. Can you think up an example of the bias that can appear there?

>  Even if using your approach, you should still keep track of the points in terms of coordinates and boxes, so that you can instantly get the boxes around. This is what Bridson says in his (very short) article.

Yeah, this is exactly what I do in my code. The problem is, the number of boxes to check (aka the size of the neighbourhood) grows exponentially with the number of dimensions (from a line to a square to a cube etc.). Here is a link to a bit more detailed discussion: https://cs.stackexchange.com/questions/109933/complexity-of-fast-poisson-disk-sampling-in-arbitrary-dimensions/135019

Reply:
> In my case, I fix it to the minimal radius r, to obtain a valid PD distribution, while also picking as many points as possible. Can you think up an example of the bias that can appear there?

If you fix the radius to the minimal radius, then you will see a clear path between all points. As the method is iterative, you will be able to draw a tree with edges of the same length.

> Yeah, this is exactly what I do in my code. The problem is, the number of boxes to check (aka the size of the neighbourhood) grows exponentially with the number of dimensions (from a line to a square to a cube etc.). Here is a link to a bit more detailed discussion: https://cs.stackexchange.com/questions/109933/complexity-of-fast-poisson-disk-sampling-in-arbitrary-dimensions/135019

OK then if we understand each other correctly it seems good and we will have to look at how we can optimize the operations here. For instance, using sets to keep track of the indices of boxes might speed up things as sets operations are really fast. And to sample a new point, I would sample a box instead of coordinates.

In any case, the next step would be to replace the code which is copyrighted (or add the written consent from the author in the comments. I just got that for the prime number code we talked about above. We could be lucky here too.).

Reply:
So far, there's no reply from the Matlab code author (after 15 years I wouldn't really expect it). The next step would be then to add the hypercube sampler and to remove the hypersphere one. I'll try to do it next week.

Reply:
Thanks for the information!

Reply:
I think we can rewrite this code anyways based on https://baezortega.github.io/2018/10/14/hypersphere-sampling/ it seems like there is some performance updates possible. If you folks agree we can give it a go still using hypersphere. You can't license formulas anyways.

Reply:
Hi Ilayn, it's a very nice and clear article. It also presents a different algorithm for sampling (without a gamma function), which makes it much easier to understand than the Matlab solution (at least for me :) ), and it can also be naturally adapted to exclude the sphere's centre. 

I suggest we indeed give it a go, and see if we can get good performance out of it.

Reply:
We can give it a go. I still thing though that at first we should go with hypercubes to be closer to the original article and maybe have a method parameter to have a hypersphere. As we said, hypersphere would be for lower dimensions as costly.

Reply:
Yes once we are done with the essential part of the algo which I don't know much about, then we can go for increasing performance, cython, pythran etc. 

Pinging @baezortega in case they are around and want to chime in because why not 😉 ?

Reply:
@tupui if you mean Bridson's Siggraph article, he doesn't use hypercubes for sampling, he uses hyperspheres. Hypercubes are used for neighbour detection, he basically splits the space with a square grid and looks in the closest cubes for overlapping discs (I use the same strategy). So hyperspheres are actually closer to the original article.

You're still right about the performance issue, hypercubes can be a nice optimisation for higher dimensions if we don't care about isotropy

Reply:
For reference, we have the go for the SO code from the author https://stackoverflow.com/questions/66047540/fast-submatrix-lookup-with-numpy#comment119793431_66054661
Thanks @diregoblin!